// @genesislcap/pbc-notify-ui/dist/esm/components/foundation-notification-dashboard/components/templates/template-dialog/template-dialog.js
import { __awaiter, __decorate } from "tslib";
import { customElement, GenesisElement, observable, volatile } from '@genesislcap/web-core';
import { Assignment, defaultUpdateType, Severity, } from '../../../../../notify.types';
import { NotifyService } from '../../../../../services/notify.service';
import { SystemService } from '../../../../../services/system.service';
import { TemplateService } from '../../../../../services/template.service';
import { humanize, logger } from '../../../../../utils';
import { DynamicRuleUtils, showNotificationError, TemplateUtils, isEmpty, } from '../../../notification-dashboard.utils';
import { DynamicRuleDialogStyles } from '../../../styles/dynamic-rule.styles';
import { TemplateDialogTemplate } from './template-dialog.template';
import { TemplateDialogMode } from './template-dialog.types';
let TemplateDialog = class TemplateDialog extends GenesisElement {
    constructor() {
        super(...arguments);
        this.templateDialogMode = null;
        this.resource = ''; // TODO: Tables + Views
        this.updateType = defaultUpdateType;
        this.parameters = [];
        this.conditions = [];
        this.resources = [];
        this.fields = [];
        this.topics = [];
        this.requiredFields = () => [this.name, this.description, this.header, this.message];
    }
    openDialog(params) {
        return __awaiter(this, void 0, void 0, function* () {
            this.topics = yield this.notifyService.getNotifyRouteTopics();
            this.resources = yield this.systemService.getResources(); // TODO: Tables + Views
            this.templateDialogMode = params.mode;
            switch (params.mode) {
                case TemplateDialogMode.CREATE:
                    yield this.createTemplate();
                    break;
                case TemplateDialogMode.EDIT:
                    yield this.editTemplate(params.data);
                    break;
                default:
                    logger.error('Error on TemplateDialogMode');
                    break;
            }
            this.dialog.show();
        });
    }
    close() {
        // Objects
        this.templateId = null;
        this.name = null;
        this.description = null;
        this.header = null;
        this.message = null;
        this.resource = null;
        this.topic = null;
        this.severity = null;
        this.templateDialogMode = null;
        // Arrays
        this.resources = [];
        this.topics = [];
        this.conditions = [];
        this.parameters = [];
        this.updateType = [];
        this.dialog.close();
        this.$emit('close');
    }
    createTemplate() {
        return __awaiter(this, void 0, void 0, function* () {
            this.conditions.push(DynamicRuleUtils.createEmptyCondition());
            this.resource = this.resources[0];
            this.fields = yield this.systemService.getFields(this.resource);
            this.topic = this.topics[0];
            this.severity = Severity.INFORMATION;
            this.updateType = defaultUpdateType;
        });
    }
    editTemplate(data) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            const template = data;
            this.templateId = template.DYNAMIC_RULE_ID;
            this.name = template.RULE_NAME;
            this.description = template.RULE_DESCRIPTION;
            this.resource = template.RULE_TABLE;
            this.updateType = template.TABLE_OPERATIONS;
            this.fields = yield this.systemService.getFields(this.resource);
            this.parameters = TemplateUtils.getParameters(template.PARAMETERS);
            this.conditions = TemplateUtils.getConditions(template.RULE_EXPRESSION.CONDITIONS, this.parameters);
            this.attributeAssignments((_a = template.RESULT_EXPRESSION) === null || _a === void 0 ? void 0 : _a.ASSIGNMENTS);
        });
    }
    attributeAssignments(assignments) {
        var _a;
        if (!assignments || !assignments.length) {
            return;
        }
        this.header = DynamicRuleUtils.getAssignmentValue(assignments, Assignment.HEADER);
        this.message = DynamicRuleUtils.getAssignmentValue(assignments, Assignment.BODY);
        this.severity = (_a = DynamicRuleUtils.getAssignmentValue(assignments, Assignment.NOTIFY_SEVERITY)) === null || _a === void 0 ? void 0 : _a.toUpperCase();
        this.topic = DynamicRuleUtils.getAssignmentValue(assignments, Assignment.TOPIC);
    }
    get templateDialogTitle() {
        return `${this.templateDialogMode === TemplateDialogMode.CREATE ? 'Create' : 'Edit'} Template`;
    }
    submit() {
        if (this.requiredFields().some((value) => isEmpty(value))) {
            return;
        }
        const template = this.createTemplateObject();
        switch (this.templateDialogMode) {
            case TemplateDialogMode.CREATE:
                this.templateService
                    .createRuleTemplate(template)
                    .then((response) => this.validateResponse(response));
                break;
            case TemplateDialogMode.EDIT:
                this.templateService
                    .editRuleTemplate(template)
                    .then((response) => this.validateResponse(response));
                break;
            default:
                logger.error('Error on TemplateDialogMode');
                break;
        }
    }
    createTemplateObject() {
        return Object.assign(Object.assign({}, (this.templateDialogMode === TemplateDialogMode.EDIT && {
            DYNAMIC_RULE_ID: this.templateId,
        })), { RULE_NAME: this.name, RULE_DESCRIPTION: this.description, RULE_TABLE: this.resource, TABLE_OPERATIONS: this.updateType, RULE_EXPRESSION: {
                CONDITIONS: DynamicRuleUtils.createConditions(this.conditions, this.parameters),
            }, RAW_RULE_EXPRESSION: null, RESULT_EXPRESSION: {
                ASSIGNMENTS: [
                    DynamicRuleUtils.createAssignment(Assignment.TOPIC, this.topic),
                    DynamicRuleUtils.createAssignment(Assignment.HEADER, this.header),
                    DynamicRuleUtils.createAssignment(Assignment.BODY, this.message),
                    DynamicRuleUtils.createAssignment(Assignment.NOTIFY_SEVERITY, humanize(this.severity)),
                ],
            }, PARAMETER_DETAILS: TemplateUtils.createParameterDetails(this.parameters) });
    }
    validateResponse(response) {
        if (response.MESSAGE_TYPE === 'EVENT_ACK') {
            this.close();
            return;
        }
        showNotificationError(response.ERROR);
    }
    resourceChanged() {
        return __awaiter(this, void 0, void 0, function* () {
            this.conditions = [DynamicRuleUtils.createEmptyCondition()];
            this.fields = yield this.systemService.getFields(this.resource);
        });
    }
    // #region Condition
    newCondition() {
        this.conditions.push(DynamicRuleUtils.createEmptyCondition());
    }
    editCondition(editedCondition) { }
    deleteCondition(deletedCondition) {
        if (this.conditions.length <= 1) {
            return;
        }
        this.conditions = this.conditions.filter((condition) => condition !== deletedCondition);
    }
    // #endregion
    // #region Parameter
    newParameter() {
        this.parameters = [...this.parameters, TemplateUtils.createEmptyParameter()];
    }
    editParameter(editedParameter) {
        this.parameters = [...this.parameters];
    }
    deleteParameter(deletedParameter) {
        if (this.conditions.find((condition) => condition.RIGHT_VALUE === deletedParameter.UUID)) {
            return;
        }
        this.parameters = this.parameters.filter((parameter) => parameter.UUID !== deletedParameter.UUID);
    }
    // #endregion
    get validateRequiredFields() {
        return this.requiredFields().some((value) => isEmpty(value));
    }
};
__decorate([
    TemplateService
], TemplateDialog.prototype, "templateService", void 0);
__decorate([
    SystemService
], TemplateDialog.prototype, "systemService", void 0);
__decorate([
    NotifyService
], TemplateDialog.prototype, "notifyService", void 0);
__decorate([
    observable
], TemplateDialog.prototype, "templateDialogMode", void 0);
__decorate([
    observable
], TemplateDialog.prototype, "templateId", void 0);
__decorate([
    observable
], TemplateDialog.prototype, "name", void 0);
__decorate([
    observable
], TemplateDialog.prototype, "description", void 0);
__decorate([
    observable
], TemplateDialog.prototype, "header", void 0);
__decorate([
    observable
], TemplateDialog.prototype, "message", void 0);
__decorate([
    observable
], TemplateDialog.prototype, "resource", void 0);
__decorate([
    observable
], TemplateDialog.prototype, "topic", void 0);
__decorate([
    observable
], TemplateDialog.prototype, "severity", void 0);
__decorate([
    observable
], TemplateDialog.prototype, "updateType", void 0);
__decorate([
    observable
], TemplateDialog.prototype, "parameters", void 0);
__decorate([
    observable
], TemplateDialog.prototype, "conditions", void 0);
__decorate([
    observable
], TemplateDialog.prototype, "resources", void 0);
__decorate([
    observable
], TemplateDialog.prototype, "fields", void 0);
__decorate([
    observable
], TemplateDialog.prototype, "topics", void 0);
__decorate([
    volatile
], TemplateDialog.prototype, "validateRequiredFields", null);
TemplateDialog = __decorate([
    customElement({
        name: 'template-dialog',
        template: TemplateDialogTemplate,
        styles: DynamicRuleDialogStyles,
    })
], TemplateDialog);
export { TemplateDialog };
