// @genesislcap/pbc-notify-ui/dist/esm/components/foundation-notification-dashboard/components/templates/template-dialog/parameter-builder/parameter-builder.js
import { __awaiter, __decorate } from "tslib";
import { attr, customElement, GenesisElement, observable } from '@genesislcap/web-core';
import { PARAM_SOURCE_TYPE_DEFINED_GROUP, PARAM_SOURCE_TYPE_REQ_REP, } from '../../../../types/param-source-type';
import { ParameterBuilderStyles as styles } from './parameter-builder.styles';
import { ParameterBuilderTemplate as template } from './parameter-builder.template';
const name = 'parameter-builder';
let ParameterBuilder = class ParameterBuilder extends GenesisElement {
    constructor() {
        super(...arguments);
        this.reqRepItems = ['REQUEST_1', 'REQUEST_2', 'REQUEST_3'];
    }
    connectedCallback() {
        const _super = Object.create(null, {
            connectedCallback: { get: () => super.connectedCallback }
        });
        return __awaiter(this, void 0, void 0, function* () {
            _super.connectedCallback.call(this);
        });
    }
    disconnectedCallback() {
        super.disconnectedCallback();
    }
    deleteParameter() {
        this.$emit('delete', this.parameter);
    }
    // #region Changed Parameters
    parameterChanged() {
        var _a, _b;
        this.name = this.parameter.PARAM_NAME;
        this.label = this.parameter.PARAM_LABEL;
        this.type = this.parameter.PARAM_TYPE;
        const paramTemp = this.parameter.PARAM_SOURCE; // Can be cleaned
        this.sourceType = this.parameter.PARAM_SOURCE_TYPE;
        switch (this.sourceType) {
            case PARAM_SOURCE_TYPE_REQ_REP:
                this.sourceSelected = (_a = this.parameter.PARAM_SOURCE) !== null && _a !== void 0 ? _a : paramTemp;
                break;
            case PARAM_SOURCE_TYPE_DEFINED_GROUP:
                this.sourceTyped = (_b = this.parameter.PARAM_SOURCE) !== null && _b !== void 0 ? _b : paramTemp;
                break;
            default:
                break;
        }
    }
    nameChanged() {
        this.genericChanged('PARAM_NAME', this.name);
    }
    labelChanged() {
        this.genericChanged('PARAM_LABEL', this.label);
    }
    typeChanged() {
        this.genericChanged('PARAM_TYPE', this.type);
    }
    sourceTypeChanged() {
        this.genericChanged('PARAM_SOURCE_TYPE', this.sourceType);
        switch (this.sourceType) {
            case PARAM_SOURCE_TYPE_REQ_REP:
                this.sourceTyped = '';
                break;
            case PARAM_SOURCE_TYPE_DEFINED_GROUP:
                this.sourceSelected = this.reqRepItems[0];
                break;
            default:
                this.sourceSelected = this.reqRepItems[0];
                this.sourceTyped = '';
                break;
        }
    }
    sourceSelectedChanged() {
        if (this.sourceType !== PARAM_SOURCE_TYPE_REQ_REP) {
            return;
        }
        this.genericChanged('PARAM_SOURCE', this.sourceSelected);
    }
    sourceTypedChanged() {
        if (this.sourceType !== PARAM_SOURCE_TYPE_DEFINED_GROUP) {
            return;
        }
        this.genericChanged('PARAM_SOURCE', this.sourceTyped);
    }
    // #endregion
    genericChanged(parameterAttribute, value) {
        if (!this.parameter) {
            return;
        }
        if (this.parameter[parameterAttribute] === value) {
            return;
        }
        this.parameter[parameterAttribute] = value;
        this.$emit('edit', this.parameter);
    }
};
__decorate([
    attr
], ParameterBuilder.prototype, "parameter", void 0);
__decorate([
    observable
], ParameterBuilder.prototype, "name", void 0);
__decorate([
    observable
], ParameterBuilder.prototype, "label", void 0);
__decorate([
    observable
], ParameterBuilder.prototype, "type", void 0);
__decorate([
    observable
], ParameterBuilder.prototype, "sourceType", void 0);
__decorate([
    observable
], ParameterBuilder.prototype, "sourceSelected", void 0);
__decorate([
    observable
], ParameterBuilder.prototype, "sourceTyped", void 0);
ParameterBuilder = __decorate([
    customElement({
        name,
        template,
        styles,
    })
], ParameterBuilder);
export { ParameterBuilder };
