// @genesislcap/pbc-notify-ui/dist/esm/components/foundation-notification-dashboard/components/templates/template-dialog/template-condition-builder/template-condition-builder.js
import { __awaiter, __decorate } from "tslib";
import { attr, customElement, GenesisElement, observable } from '@genesislcap/web-core';
import { LogicalOperator, RightCriteria, nullAndBlankLogicalOperatorValues, } from '../../../../../../notify.types';
import { logger } from '../../../../../../utils';
import { ConditionBuilderStyles } from '../../../../styles/condition-builder.styles';
import { TemplateConditionBuilderTemplate } from './template-condition-builder.template';
let TemplateConditionBuilder = class TemplateConditionBuilder extends GenesisElement {
    constructor() {
        super(...arguments);
        this.leftValueItems = [];
        this.rightCriteria = RightCriteria.VALUE;
        this.rightValueItems = [];
        // #endregion
    }
    deleteCondition() {
        this.$emit('delete', this.condition);
    }
    conditionChanged() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.fields) {
                this.leftValueItems = this.fields;
            }
            this.assignLeftValue();
            this.logicalOperator = this.condition.LOGICAL_OPERATOR;
            const rightValue = this.condition.RIGHT_VALUE;
            this.rightCriteria = this.condition.RIGHT_CRITERIA;
            this.condition.RIGHT_VALUE = rightValue;
            this.assignRightValue();
            this.$emit('edit', this.condition);
        });
    }
    parametersChanged() {
        if (!Array.isArray(this.parameters) || !this.parameters) {
            return;
        }
        this.rightValueItems = this.parameters;
    }
    fieldsChanged() {
        if (!Array.isArray(this.fields) || !this.fields.length) {
            return;
        }
        this.leftValueItems = this.fields;
        this.leftValue = JSON.stringify(this.leftValueItems[0]);
    }
    logicalOperatorChanged() {
        if (!this.condition) {
            return;
        }
        if (nullAndBlankLogicalOperatorValues.includes(LogicalOperator[this.logicalOperator])) {
            this.rightCriteria = RightCriteria.VALUE;
            this.clearRightValueText();
        }
        this.condition.LOGICAL_OPERATOR = this.logicalOperator;
        this.$emit('edit', this.condition);
    }
    // #region LeftValue
    assignLeftValue() {
        // Add (fresh new; assign the 1st element)
        if (!this.condition.LEFT_VALUE) {
            this.leftValue = JSON.stringify(this.leftValueItems[0]);
            return;
        }
        // Add (value changed)
        if (typeof this.condition.LEFT_VALUE !== 'string') {
            this.leftValue = JSON.stringify(this.condition.LEFT_VALUE);
            return;
        }
        // Edit
        this.leftValue = JSON.stringify(this.leftValueItems.find((item) => item.FIELD_NAME === this.condition.LEFT_VALUE));
    }
    leftValueChanged() {
        if (!this.condition) {
            return;
        }
        this.condition.LEFT_VALUE = JSON.parse(this.leftValue);
        this.$emit('edit', this.condition);
    }
    // #endregion
    // #region RightValue
    assignRightValue() {
        var _a;
        if (this.condition.RIGHT_CRITERIA === RightCriteria.VALUE) {
            this.rightValueText = this.condition.RIGHT_VALUE;
            return;
        }
        if (((_a = this.parameters) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            this.rightValueSelect = this.condition.RIGHT_VALUE;
            return;
        }
        this.rightValueSelect = null;
    }
    rightCriteriaChanged() {
        if (!this.condition) {
            return;
        }
        this.condition.RIGHT_CRITERIA = this.rightCriteria;
        switch (this.rightCriteria) {
            case RightCriteria.VALUE:
                this.clearRightValueText();
                break;
            case RightCriteria.USER_ENTRY:
                this.clearRightValueSelect();
                break;
            default:
                logger.error(`Unknown RightCriteria: ${this.rightCriteria}`);
                break;
        }
        this.$emit('edit', this.condition);
    }
    clearRightValueText() {
        this.rightValueText = '';
    }
    clearRightValueSelect() {
        var _a;
        this.rightValueSelect = ((_a = this.parameters) === null || _a === void 0 ? void 0 : _a.length) > 0 ? this.parameters[0].UUID : null;
    }
    rightValueTextChanged() {
        if (!this.condition) {
            return;
        }
        this.condition.RIGHT_VALUE = this.rightValueText;
        this.$emit('edit', this.condition);
    }
    rightValueSelectChanged() {
        if (!this.condition) {
            return;
        }
        this.condition.RIGHT_VALUE = this.rightValueSelect;
        this.$emit('edit', this.condition);
    }
};
__decorate([
    attr
], TemplateConditionBuilder.prototype, "condition", void 0);
__decorate([
    attr
], TemplateConditionBuilder.prototype, "parameters", void 0);
__decorate([
    attr
], TemplateConditionBuilder.prototype, "fields", void 0);
__decorate([
    observable
], TemplateConditionBuilder.prototype, "leftValue", void 0);
__decorate([
    observable
], TemplateConditionBuilder.prototype, "leftValueItems", void 0);
__decorate([
    observable
], TemplateConditionBuilder.prototype, "logicalOperator", void 0);
__decorate([
    observable
], TemplateConditionBuilder.prototype, "rightCriteria", void 0);
__decorate([
    observable
], TemplateConditionBuilder.prototype, "rightValueText", void 0);
__decorate([
    observable
], TemplateConditionBuilder.prototype, "rightValueSelect", void 0);
__decorate([
    observable
], TemplateConditionBuilder.prototype, "rightValueItems", void 0);
TemplateConditionBuilder = __decorate([
    customElement({
        name: 'template-condition-builder',
        template: TemplateConditionBuilderTemplate,
        styles: ConditionBuilderStyles,
    })
], TemplateConditionBuilder);
export { TemplateConditionBuilder };
