// @genesislcap/pbc-notify-ui/dist/esm/components/foundation-notification-dashboard/components/rules/rule-dialog/rule-dialog.js
import { __awaiter, __decorate } from "tslib";
import { customElement, GenesisElement, observable, volatile } from '@genesislcap/web-core';
import { Assignment, defaultUpdateType, Severity, } from '../../../../../notify.types';
import { NotifyService } from '../../../../../services/notify.service';
import { RuleService } from '../../../../../services/rule.service';
import { SystemService } from '../../../../../services/system.service';
import { humanize } from '../../../../../utils';
import { DynamicRuleUtils, RuleUtils, showNotificationError, isEmpty, } from '../../../notification-dashboard.utils';
import { DynamicRuleDialogStyles } from '../../../styles/dynamic-rule.styles';
import { RuleDialogTemplate } from './rule-dialog.template';
import { RuleDialogMode } from './rule-dialog.types';
let RuleDialog = class RuleDialog extends GenesisElement {
    constructor() {
        super(...arguments);
        this.resource = ''; // TODO: Tables + Views
        this.topic = '';
        this.severity = '';
        this.updateType = defaultUpdateType;
        this.conditions = [];
        this.resources = [];
        this.fields = [];
        this.requiredFields = () => [this.name, this.description, this.header, this.message];
    }
    openDialog(params) {
        return __awaiter(this, void 0, void 0, function* () {
            this.topics = yield this.notifyService.getNotifyRouteTopics();
            this.resources = yield this.systemService.getResources();
            this.ruleDialogMode = params.mode;
            switch (params.mode) {
                case RuleDialogMode.CREATE:
                    yield this.createRule();
                    break;
                case RuleDialogMode.EDIT:
                    yield this.editRule(params.data);
                    break;
            }
            this.dialog.show();
        });
    }
    close() {
        // Objects
        this.ruleId = null;
        this.name = null;
        this.description = null;
        this.header = null;
        this.message = null;
        this.resource = null;
        this.topic = null;
        this.severity = null;
        this.ruleDialogMode = null;
        // Arrays
        this.resources = [];
        this.topics = [];
        this.conditions = [];
        this.updateType = [];
        this.dialog.close();
        this.$emit('close');
    }
    createRule() {
        return __awaiter(this, void 0, void 0, function* () {
            this.conditions.push(DynamicRuleUtils.createEmptyCondition());
            this.resource = this.resources[0];
            this.fields = yield this.systemService.getFields(this.resource);
            this.topic = this.topics[0];
            this.severity = Severity.INFORMATION;
            this.updateType = defaultUpdateType;
        });
    }
    editRule(data) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            const rule = data;
            this.ruleId = rule.DYNAMIC_RULE_ID;
            this.name = rule.RULE_NAME;
            this.description = rule.RULE_DESCRIPTION;
            this.resource = rule.RULE_TABLE;
            this.updateType = rule.TABLE_OPERATIONS;
            this.fields = yield this.systemService.getFields(this.resource);
            this.conditions = RuleUtils.getConditions(rule.RULE_EXPRESSION.CONDITIONS);
            this.attributeAssignments((_a = rule.RESULT_EXPRESSION) === null || _a === void 0 ? void 0 : _a.ASSIGNMENTS);
        });
    }
    attributeAssignments(assignments) {
        var _a;
        if (!assignments || !assignments.length) {
            return;
        }
        this.header = DynamicRuleUtils.getAssignmentValue(assignments, Assignment.HEADER);
        this.message = DynamicRuleUtils.getAssignmentValue(assignments, Assignment.BODY);
        this.severity = (_a = DynamicRuleUtils.getAssignmentValue(assignments, Assignment.NOTIFY_SEVERITY)) === null || _a === void 0 ? void 0 : _a.toUpperCase();
        this.topic = DynamicRuleUtils.getAssignmentValue(assignments, Assignment.TOPIC);
    }
    get ruleDialogTitle() {
        return `${this.ruleDialogMode === RuleDialogMode.CREATE ? 'Create' : 'Edit'} Rule`;
    }
    submit() {
        if (this.requiredFields().some((value) => isEmpty(value))) {
            return;
        }
        const rule = this.createRuleObject();
        switch (this.ruleDialogMode) {
            case RuleDialogMode.CREATE:
                this.ruleService.createRule(rule).then((response) => this.validateResponse(response));
                break;
            case RuleDialogMode.EDIT:
                this.ruleService.updateRule(rule).then((response) => this.validateResponse(response));
                break;
        }
    }
    createRuleObject() {
        return Object.assign(Object.assign({}, (this.ruleDialogMode === RuleDialogMode.EDIT && { DYNAMIC_RULE_ID: this.ruleId })), { RULE_NAME: this.name, RULE_DESCRIPTION: this.description, RULE_TABLE: this.resource, TABLE_OPERATIONS: this.updateType, RULE_EXPRESSION: {
                CONDITIONS: DynamicRuleUtils.createConditions(this.conditions),
            }, RESULT_EXPRESSION: {
                ASSIGNMENTS: [
                    DynamicRuleUtils.createAssignment(Assignment.TOPIC, this.topic),
                    DynamicRuleUtils.createAssignment(Assignment.HEADER, this.header),
                    DynamicRuleUtils.createAssignment(Assignment.BODY, this.message),
                    DynamicRuleUtils.createAssignment(Assignment.NOTIFY_SEVERITY, humanize(this.severity)),
                ],
            } });
    }
    validateResponse(response) {
        if (response.MESSAGE_TYPE === 'EVENT_ACK') {
            this.close();
            return;
        }
        showNotificationError(response.ERROR);
    }
    resourceChanged() {
        return __awaiter(this, void 0, void 0, function* () {
            this.conditions = [DynamicRuleUtils.createEmptyCondition()];
            this.fields = yield this.systemService.getFields(this.resource);
        });
    }
    // #region Condition
    newCondition() {
        this.conditions.push(DynamicRuleUtils.createEmptyCondition());
    }
    editCondition(editedCondition) { }
    deleteCondition(deletedCondition) {
        if (this.conditions.length <= 1) {
            return;
        }
        this.conditions = this.conditions.filter((condition) => condition !== deletedCondition);
    }
    // #endregion
    get validateRequiredFields() {
        return this.requiredFields().some((value) => isEmpty(value));
    }
};
__decorate([
    RuleService
], RuleDialog.prototype, "ruleService", void 0);
__decorate([
    SystemService
], RuleDialog.prototype, "systemService", void 0);
__decorate([
    NotifyService
], RuleDialog.prototype, "notifyService", void 0);
__decorate([
    observable
], RuleDialog.prototype, "ruleDialogMode", void 0);
__decorate([
    observable
], RuleDialog.prototype, "name", void 0);
__decorate([
    observable
], RuleDialog.prototype, "description", void 0);
__decorate([
    observable
], RuleDialog.prototype, "resource", void 0);
__decorate([
    observable
], RuleDialog.prototype, "topic", void 0);
__decorate([
    observable
], RuleDialog.prototype, "severity", void 0);
__decorate([
    observable
], RuleDialog.prototype, "updateType", void 0);
__decorate([
    observable
], RuleDialog.prototype, "header", void 0);
__decorate([
    observable
], RuleDialog.prototype, "message", void 0);
__decorate([
    observable
], RuleDialog.prototype, "conditions", void 0);
__decorate([
    observable
], RuleDialog.prototype, "resources", void 0);
__decorate([
    observable
], RuleDialog.prototype, "fields", void 0);
__decorate([
    observable
], RuleDialog.prototype, "topics", void 0);
__decorate([
    volatile
], RuleDialog.prototype, "validateRequiredFields", null);
RuleDialog = __decorate([
    customElement({
        name: 'rule-dialog',
        template: RuleDialogTemplate,
        styles: DynamicRuleDialogStyles,
    })
], RuleDialog);
export { RuleDialog };
