// @genesislcap/pbc-notify-ui/dist/esm/components/foundation-notification-dashboard/components/rules/rule-dialog/rule-condition-builder/rule-condition-builder.js
import { __awaiter, __decorate } from "tslib";
import { customElement, GenesisElement, observable } from '@genesislcap/web-core';
import { LogicalOperator, nullAndBlankLogicalOperatorValues, } from '../../../../../../notify.types';
import { logger } from '../../../../../../utils';
import { ConditionBuilderStyles } from '../../../../styles/condition-builder.styles';
import { RuleConditionBuilderTemplate } from './rule-condition-builder.template';
import { RightCriteria } from './rule-condition-builder.types';
let RuleConditionBuilder = class RuleConditionBuilder extends GenesisElement {
    constructor() {
        super(...arguments);
        this.leftValueItems = [];
        this.rightCriteria = RightCriteria.VALUE;
    }
    deleteCondition() {
        this.$emit('delete', this.condition);
    }
    conditionChanged() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.fields) {
                this.leftValueItems = this.fields;
            }
            this.assignLeftValue();
            this.logicalOperator = this.condition.LOGICAL_OPERATOR;
            this.rightCriteria = this.condition.RIGHT_CRITERIA;
            this.assignRightValue();
            this.$emit('edit', this.condition);
        });
    }
    fieldsChanged() {
        if (!Array.isArray(this.fields) || !this.fields.length) {
            return;
        }
        this.leftValueItems = this.fields;
        this.leftValue = JSON.stringify(this.leftValueItems[0]);
    }
    assignLeftValue() {
        // Add (fresh new; assign the 1st element)
        if (!this.condition.LEFT_VALUE) {
            this.leftValue = JSON.stringify(this.leftValueItems[0]);
            return;
        }
        // Add (value changed)
        if (typeof this.condition.LEFT_VALUE !== 'string') {
            this.leftValue = JSON.stringify(this.condition.LEFT_VALUE);
            return;
        }
        // Edit
        this.leftValue = JSON.stringify(this.leftValueItems.find((item) => item.FIELD_NAME === this.condition.LEFT_VALUE));
    }
    assignRightValue() {
        if (this.condition.RIGHT_CRITERIA === RightCriteria.VALUE) {
            this.rightValueText = this.condition.RIGHT_VALUE;
            return;
        }
        this.rightValueSelect = null;
    }
    leftValueChanged() {
        if (!this.condition) {
            return;
        }
        this.condition.LEFT_VALUE = JSON.parse(this.leftValue);
        this.$emit('edit', this.condition);
    }
    logicalOperatorChanged() {
        if (!this.condition) {
            return;
        }
        if (nullAndBlankLogicalOperatorValues.includes(LogicalOperator[this.logicalOperator])) {
            this.rightCriteria = RightCriteria.VALUE;
            this.clearRightValueText();
        }
        this.condition.LOGICAL_OPERATOR = this.logicalOperator;
        this.$emit('edit', this.condition);
    }
    rightCriteriaChanged() {
        if (!this.condition) {
            return;
        }
        this.condition.RIGHT_CRITERIA = this.rightCriteria;
        switch (this.rightCriteria) {
            case RightCriteria.VALUE:
                this.clearRightValueText();
                break;
            case RightCriteria.FIELD:
                this.clearRightValueSelect();
                break;
            default:
                logger.error(`Unknown RightCriteria: ${this.rightCriteria}`);
                break;
        }
        this.$emit('edit', this.condition);
    }
    clearRightValueText() {
        this.rightValueText = '';
    }
    clearRightValueSelect() {
        this.rightValueSelect = this.fields[0].FIELD_NAME;
    }
    rightValueTextChanged() {
        if (!this.condition) {
            return;
        }
        this.condition.RIGHT_VALUE = this.rightValueText;
        this.$emit('edit', this.condition);
    }
    rightValueSelectChanged() {
        if (!this.condition) {
            return;
        }
        this.condition.RIGHT_VALUE = this.rightValueSelect;
        this.$emit('edit', this.condition);
    }
};
__decorate([
    observable
], RuleConditionBuilder.prototype, "condition", void 0);
__decorate([
    observable
], RuleConditionBuilder.prototype, "fields", void 0);
__decorate([
    observable
], RuleConditionBuilder.prototype, "leftValue", void 0);
__decorate([
    observable
], RuleConditionBuilder.prototype, "leftValueItems", void 0);
__decorate([
    observable
], RuleConditionBuilder.prototype, "logicalOperator", void 0);
__decorate([
    observable
], RuleConditionBuilder.prototype, "rightCriteria", void 0);
__decorate([
    observable
], RuleConditionBuilder.prototype, "rightValueText", void 0);
__decorate([
    observable
], RuleConditionBuilder.prototype, "rightValueSelect", void 0);
RuleConditionBuilder = __decorate([
    customElement({
        name: 'rule-condition-builder',
        template: RuleConditionBuilderTemplate,
        styles: ConditionBuilderStyles,
    })
], RuleConditionBuilder);
export { RuleConditionBuilder };
