// @genesislcap/grid-pro/dist/esm/state-persistence/kv-state.js
import { __awaiter, __decorate } from "tslib";
import { KVStorage, Session } from '@genesislcap/foundation-comms';
import { logger } from '../utils';
/**
 * Implementation for the KV Storage State Persistence interface.
 * @public
 */
export class KVStorageStatePersistence {
    getColumnState(persistColumnStateKey) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            yield this.migrateLocalStorageToKVStorage(persistColumnStateKey);
            const kvColumnState = yield this.kvStorage.get(persistColumnStateKey);
            const storageColumnState = kvColumnState ? (_b = (_a = kvColumnState[0]) === null || _a === void 0 ? void 0 : _a.kv) === null || _b === void 0 ? void 0 : _b.value : undefined;
            if (storageColumnState) {
                return JSON.parse(storageColumnState);
            }
            return [];
        });
    }
    saveColumnState(persistColumnStateKey, columnState) {
        return __awaiter(this, void 0, void 0, function* () {
            this.kvStorage.put([
                {
                    key: persistColumnStateKey,
                    value: JSON.stringify(columnState),
                },
            ]);
        });
    }
    migrateLocalStorageToKVStorage(persistColumnStateKey) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            const kv = yield this.kvStorage.get(persistColumnStateKey);
            if (kv && ((_b = (_a = kv[0]) === null || _a === void 0 ? void 0 : _a.kv) === null || _b === void 0 ? void 0 : _b.value))
                return;
            const localGridOptions = this.session.getLocalStorageItem('gridOptions');
            const localState = localGridOptions
                ? JSON.parse(localGridOptions)[persistColumnStateKey]
                : undefined;
            if (localState) {
                logger.debug(`Columns state migration for key: ${persistColumnStateKey} from LocalStorage to KV Storage`);
                this.kvStorage.put([
                    {
                        key: persistColumnStateKey,
                        value: localState,
                    },
                ]);
            }
        });
    }
    getFilterModel(persistFilterModelKey) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            const kvFilterModel = yield this.kvStorage.get(persistFilterModelKey);
            const storagetFilterModel = kvFilterModel ? (_b = (_a = kvFilterModel[0]) === null || _a === void 0 ? void 0 : _a.kv) === null || _b === void 0 ? void 0 : _b.value : undefined;
            if (storagetFilterModel) {
                return JSON.parse(storagetFilterModel);
            }
            return [];
        });
    }
    saveFilterModel(persistFilterModelKey, filterModel) {
        return __awaiter(this, void 0, void 0, function* () {
            this.kvStorage.put([
                {
                    key: persistFilterModelKey,
                    value: JSON.stringify(filterModel),
                },
            ]);
        });
    }
}
__decorate([
    KVStorage
], KVStorageStatePersistence.prototype, "kvStorage", void 0);
__decorate([
    Session
], KVStorageStatePersistence.prototype, "session", void 0);
