// @genesislcap/pbc-notify-ui/dist/esm/components/foundation-inbox/components/inbox-subscription/rule-parameter/rule-parameter.js
import { __decorate } from "tslib";
import { attr, customElement, GenesisElement, observable } from '@genesislcap/web-core';
import { PARAM_SOURCE_TYPE_DEFINED_GROUP, PARAM_SOURCE_TYPE_USER_TEXT, } from '../../../../foundation-notification-dashboard/types/param-source-type';
import { PARAM_TYPE_BOOLEAN } from '../../../../foundation-notification-dashboard/types/param-type';
import { RuleParameterStyles } from './rule-parameter.styles';
import { RuleParameterTemplate } from './rule-parameter.template';
let RuleParameter = class RuleParameter extends GenesisElement {
    constructor() {
        super(...arguments);
        this.isFirstElement = false;
        this.inputValue = '';
    }
    connectedCallback() {
        super.connectedCallback();
        switch (this.parameter.PARAM_SOURCE_TYPE) {
            case PARAM_SOURCE_TYPE_DEFINED_GROUP:
                this.inputValue = this.parameter.PARAM_SOURCE.split(',')[0];
                break;
            case PARAM_SOURCE_TYPE_USER_TEXT:
                switch (this.parameter.PARAM_TYPE) {
                    case PARAM_TYPE_BOOLEAN:
                        this.inputValue = 'true';
                        break;
                    default:
                        this.inputValue = '';
                        break;
                }
                break;
            default:
                break;
        }
    }
    disconnectedCallback() {
        super.disconnectedCallback();
    }
    parameterEdit(inputValue) {
        this.inputValue = inputValue;
        this.$emit('parameterEdited', {
            parameter: this.parameter,
            valueTyped: inputValue,
        });
    }
};
__decorate([
    attr,
    observable
], RuleParameter.prototype, "parameter", void 0);
__decorate([
    attr
], RuleParameter.prototype, "isFirstElement", void 0);
__decorate([
    attr
], RuleParameter.prototype, "operator", void 0);
__decorate([
    observable
], RuleParameter.prototype, "inputValue", void 0);
RuleParameter = __decorate([
    customElement({
        name: 'rule-parameter',
        template: RuleParameterTemplate,
        styles: RuleParameterStyles,
    })
], RuleParameter);
export { RuleParameter };
